File: cs/DataStructures/Stack.cs
namespace cs.DataStructures;

public class Stack
{
    private readonly int[] _stack;
    private int _top = -1;

    private Stack(int size)
    {
        _stack = new int[size];
    }

    private void Push(int value)
    {
        if (_top == _stack.Length - 1)
        {
            throw new Exception("Stack overflow");
        }

        _stack[++_top] = value;
    }
    
    private int Pop()
    {
        if (_top == -1)
        {
            throw new Exception("Stack underflow");
        }

        return _stack[_top--];
    }
    
    private int Peek()
    {
        if (_top == -1)
        {
            throw new Exception("Stack underflow");
        }

        return _stack[_top];
    }
    
    private bool IsEmpty()
    {
        return _top == -1;
    }
    
    private bool IsFull()
    {
        return _top == _stack.Length - 1;
    }

public static void Main()
    {
        var stack = new Stack(5);
        
        stack.Push(1);
        stack.Push(2);
        stack.Push(3);
        stack.Push(4);
        stack.Push(5);
        
        Console.WriteLine($"Is Full: {stack.IsFull()} ");

        stack.Pop();
        var popped = stack.Pop();
        Console.WriteLine($"Popped: {popped}");
        
        Console.WriteLine($"Is Empty: {stack.IsEmpty()}");

        Console.WriteLine($"Peeking: {stack.Peek()}");
    }
}

File: cpp/dataStructures/stack.cpp
#include <iostream>

using namespace std;

#define STACK_SIZE 100

class Stack
{
private:
    int stack[STACK_SIZE];
    int top = -1;

public:
    void push(int item)
    {
        if (top >= STACK_SIZE - 1)
        {
            cout << "Stack Overflow" << endl;
            return;
        }
        top++;
        stack[top] = item;
    }

    void pop()
    {
        if (top <= -1)
        {
            cout << "Stack Underflow" << endl;
            return;
        }
        top--;
    }

    void peek()
    {
        cout << "TOP: " << top << endl;
        cout << "Peek Top: " << stack[top] << endl;
    }

    bool isEmpty()
    {
        return top == 0;
    }

    bool isFull()
    {
        return top == STACK_SIZE;
    }

    void print()
    {
        if (top >= 0)
        {
            for (int i = 0; i <= top; i++)
            {
                cout << "Index " << i << ": " << stack[i] << endl;
            }
        }
    }
};

int main()
{
    cout << "stack implementation" << endl;

    Stack s;

    cout << "Is empty?: " << s.isEmpty() << endl;

    s.push(10);

    cout << "Is empty?: " << s.isEmpty() << endl;

    s.push(20);
    s.push(30);
    s.pop();
    s.push(40);
    s.push(50);

    s.peek();

    s.print();
};
